compiler/qsc_qasm3/src/semantic/lowerer.rs

```diff
@@ -953,8 +1016,172 @@ impl Lowerer {
+        modifiers.reverse();
+        let modifiers = list_from_iter(modifiers);
+        semantic::StmtKind::GateCall(semantic::GateCall {
+            span: stmt.span,
+            modifiers,
+            symbol_id,
+            args,
+            qubits,
+            duration,
+            quantum_arity,
+            quantum_arity_with_modifiers,
+        })
+
+        // The compiler will be left to do all things that need explicit Q# knowledge.
+        // But it won't need to check arities, know about implicit modifiers, or do
+        // any casting of classical args. There is still some inherit complexity to
+        // building a Q# gate call with this information, but it won't be cluttered
+        // by all the semantic analysis.
+    }
+
+    fn lower_modifier(
+        &mut self,
+        modifier: &syntax::QuantumGateModifier,
+    ) -> Option<semantic::QuantumGateModifier> {
+        let kind = match &modifier.kind {
+            syntax::GateModifierKind::Inv => semantic::GateModifierKind::Inv,
+            syntax::GateModifierKind::Pow(expr) => {
+                semantic::GateModifierKind::Pow(self.lower_expr(expr))
+            }
+            syntax::GateModifierKind::Ctrl(expr) => {
+                let ctrl_args = self.lower_modifier_ctrl_args(expr.as_ref())?;
+                semantic::GateModifierKind::Ctrl(ctrl_args)
+            }
+            syntax::GateModifierKind::NegCtrl(expr) => {
+                let ctrl_args = self.lower_modifier_ctrl_args(expr.as_ref())?;
+                semantic::GateModifierKind::NegCtrl(ctrl_args)
+            }
+        };
+
+        Some(semantic::QuantumGateModifier {
+            span: modifier.span,
+            kind,
+        })
+    }
+
+    fn lower_modifier_ctrl_args(&mut self, expr: Option<&syntax::Expr>) -> Option<u32> {
+        let Some(expr) = expr else {
+            return Some(1);
+        };
+
+        let expr = self.lower_expr(expr);
+
+        let target_ty = &Type::UInt(None, true);
+        let Some(expr) = self.try_cast_expr_to_type(target_ty, &expr) else {
+            self.push_invalid_cast_error(target_ty, &expr.ty, expr.span);
+            return None;
+        };
+        let Some(lit) = expr.const_eval(&self.symbols) else {
+            self.push_semantic_error(SemanticErrorKind::ExprMustBeConst(
+                "ctrl modifier argument".into(),
+                expr.span,
+            ));
+            return None;
+        };
+
+        let semantic::LiteralKind::Int(n) = lit else {
+            unreachable!("we casted the expr to UInt before const evaluating it")
+        };
+
+        let Ok(n) = u32::try_from(n) else {
+            self.push_semantic_error(SemanticErrorKind::ExprMustFitInU32(
+                "ctrl modifier argument".into(),
+                expr.span,
+            ));
+            return None;
+        };
+
+        Some(n)
     }
 
     fn lower_gphase(&mut self, stmt: &syntax::GPhase) -> semantic::StmtKind {
```

Comment:
Need to come back to this, we added some gphase infra, but we aren't lowering the stmt yet.